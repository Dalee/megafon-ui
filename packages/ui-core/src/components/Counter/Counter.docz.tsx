import * as React from 'react';

export const flexStyle = {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'space-around',
    textAlign: 'center',
};

interface IDemoCounterWrapperProps {
    children: (prop: {
        onChange: (value: number) => void;
        initialValue: number;
    }) => JSX.Element;
    initialValue: number;
}

export const DemoCounterWrapper: React.FC<IDemoCounterWrapperProps> = ({ initialValue = 0, children }) => {
    const [value, setValue] = React.useState(initialValue);

    const handleChange = (currentValue) => {
        setValue(currentValue);
    };

    return (
        <div>
            <p style={{ marginTop: 0 }}>Value: {value}</p>
            {children({
                onChange: handleChange,
                initialValue: initialValue,
            })}
        </div>
    );
};
