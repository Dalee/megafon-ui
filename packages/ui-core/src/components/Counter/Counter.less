@import '~styles/base';
@b: e('.counter');

@{b} {
    background-color: @clearWhite;
    box-sizing: border-box;
    display: flex;
    justify-content: space-between;
    height: 32px;
    width: 120px;

    &__btn {
        background-color: transparent;
        border: 1px solid @spbSky2;
        box-sizing: border-box;
        cursor: pointer;
        display: flex;
        height: 100%;
        font-size: 20px;
        flex-grow: 1;
        align-items: center;
        justify-content: center;
        outline: none;
        padding: 0;
        -webkit-appearance: none;
        -moz-appearance: none;
        -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
        text-decoration: none;
        min-width: 32px;

        &:hover {
            background-color: @spbSky0;
        }

        &:disabled {
            cursor: default;
            color: @spbSky2;
            background-color: @spbSky1;
        }
    }

    &__btn_left {
        border-top-left-radius: 50px;
        border-bottom-left-radius: 50px;
    }

    &__btn_right {
        border-top-right-radius: 50px;
        border-bottom-right-radius: 50px;
    }

    &__minus {
        position: relative;
        top: -1px;
    }

    &__plus {
        position: relative;
        top: 1px;
    }

    &__input-box {
        border-top: 1px solid @spbSky2;
        border-bottom: 1px solid @spbSky2;
        box-sizing: border-box;
        display: flex;
        align-items: center;
        overflow: hidden;
        height: 100%;
    }

    &__input {
        border: none;
        box-sizing: border-box;
        cursor: text;
        height: 100%;
        -webkit-appearance: none;
        -moz-appearance: none;
        box-shadow: none;
        overflow: auto;
        outline: none;
        margin: 0;
        padding: 0;
        text-align: center;
        width: 100%;
    }

    &_disabled {
        @{b}__btn {
            background-color: @spbSky1;
            color: @spbSky2;
            cursor: default;
        }

        @{b}__input {
            background-color: @spbSky1;
            color: rgba(51, 51, 51, 0.5);
            cursor: default;
        }
    }
}
