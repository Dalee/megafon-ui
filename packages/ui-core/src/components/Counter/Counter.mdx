---
name: Counter
menu: Components
route: components/counter
---

import { Playground, Props } from 'docz';
import Counter from './Counter';
import { DemoCounterWrapper, flexStyle } from './Counter.docz';

# Counter

## Properties

<Props of={Counter} />

```javascript
export const DemoCounterWrapper: React.FC<IDemoCounterWrapperProps> = ({ initialValue = 0, children }) => {
    const [value, setValue] = React.useState(initialValue);

    const handleChange = (value) => {
        setValue(value);
    };

    return (
        <div>
            <p style={{ marginTop: 0 }}>Value: {value}</p>
            {children({
                onChange: handleChange,
                initialValue: initialValue,
            })}
        </div>
    );
};
```

## Basic usage

<Playground>
    <div style={flexStyle}>
        <DemoCounterWrapper initialValue={4}>
            {({ onChange, initialValue }) =>
                <Counter
                    initialValue={initialValue}
                    onChange={onChange}
                    min={3}
                    max={33}
                />
            }
        </DemoCounterWrapper>
        <DemoCounterWrapper initialValue={18}>
            {({ onChange, initialValue }) =>
                <Counter
                    initialValue={initialValue}
                    onChange={onChange}
                    min={3}
                    max={33}
                />
            }
        </DemoCounterWrapper>
        <DemoCounterWrapper initialValue={33}>
            {({ onChange, initialValue }) =>
                <Counter
                    initialValue={initialValue}
                    onChange={onChange}
                    min={3}
                    max={33}
                />
            }
        </DemoCounterWrapper>
        <DemoCounterWrapper>
            {({ onChange }) =>
                <Counter
                    onChange={onChange}
                    isDisabled
                />
            }
        </DemoCounterWrapper>
    </div>
</Playground>
