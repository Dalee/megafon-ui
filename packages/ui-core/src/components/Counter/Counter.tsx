import * as React from 'react';
import cnCreate from 'utils/cnCreate';
import './Counter.less';

export interface ICounterProps {
    /** Дополнительный класс корневого элемента */
    className?: string;
    /** Начальное значение */
    initialValue?: number;
    /** Минимальное доступное значение */
    min?: number;
    /** Максимальное доступное значение */
    max?: number;
    /** Запретить взаимодействие с компонентом */
    isDisabled?: boolean;
    /** Обработчик изменения значения 'value' */
    onChange?: (value: number) => void;
}

const cn = cnCreate('mfui-beta-counter');
const Counter: React.FC<ICounterProps> = ({
    className,
    initialValue,
    max = 999999,
    min = 0,
    isDisabled = false,
    onChange,
}) => {
    const currentInitialValue = initialValue || min;
    const [counter, setCounter] = React.useState(currentInitialValue);

    React.useEffect(() => {
        setCounter(currentInitialValue);
    }, [currentInitialValue]);

    function handleValueChange(value: number) {
        setCounter(value);

        if (value < min) {
            onChange && onChange(min);
        }

        if (value > max) {
            onChange && onChange(max);
        }

        if (value >= min && value <= max) {
            onChange && onChange(value);
        }
    }

    function handleMinusClick(): void {
        handleValueChange(counter - 1);
    }

    function handlePlusClick(): void {
        handleValueChange(counter + 1);
    }

    function handleInputChange(e: React.ChangeEvent<HTMLInputElement>): void {
        const pattern = /^[0-9\b]+$/;
        const { value } = e.target;

        if (value !== '' && !pattern.test(value)) {
            return;
        }

        handleValueChange(Number(value));
    }

    function handleInputBlur(e: React.FocusEvent<HTMLInputElement>): void {
        const { value } = e.target;
        const numberValue = Number(value);

        if (numberValue < min) {
            handleValueChange(min);

            return;
        }

        if (numberValue > max) {
            handleValueChange(max);

            return;
        }
    }

    return (
        <div className={cn({ disabled: isDisabled }, className)}>
            <button
                className={cn('btn', { left: true })}
                type="button"
                disabled={isDisabled || counter <= min}
                onClick={handleMinusClick}
            >
                <span className={cn('minus')}>–</span>
            </button>
            <div className={cn('input-box')}>
                <input
                    className={cn('input')}
                    value={counter}
                    onChange={handleInputChange}
                    onBlur={handleInputBlur}
                    disabled={isDisabled}
                />
            </div>
            <button
                className={cn('btn', { right: true })}
                type="button"
                disabled={isDisabled || counter >= max}
                onClick={handlePlusClick}
            >
                <span className={cn('plus')}>+</span>
            </button>
        </div>
    );
};

export default Counter;
