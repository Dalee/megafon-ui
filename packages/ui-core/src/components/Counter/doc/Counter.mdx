---
name: Counter
menu: Компоненты
route: components/counter
---

import { Playground, Props } from 'docz';
import Counter from '../Counter';
import Example from './Counter.example.mdx';

# Counter

<ImportExample name="Counter" />

<DoczTabs>
    <Example />
    <Props of={Counter} />
</DoczTabs>
