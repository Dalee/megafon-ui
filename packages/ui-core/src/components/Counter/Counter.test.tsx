import * as React from 'react';
import { shallow } from 'enzyme';
import Counter from './Counter';

const props = {
    className: 'test-class',
    initialValue: 10,
    min: 3,
    max: 33,
    isDisabled: true,
};

describe('<Counter />', () => {
    describe('layout', () => {
        it('renders Counter', () => {
            const wrapper = shallow(
                <Counter />
            );
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('layout', () => {
        it('renders Counter with props', () => {
            const wrapper = shallow(
                <Counter {...props} />
            );
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('handlers', () => {
        it('calls onClick for left button', () => {
            const handleClick = jest.fn();

            const wrapper = shallow(
                <Counter initialValue={10} onChange={handleClick} />
            );

            wrapper.find('button').first().simulate('click');
            expect(handleClick).toHaveBeenCalledWith(9);
        });
    });

    describe('handlers', () => {
        it('calls onClick for right button', () => {
            const handleClick = jest.fn();

            const wrapper = shallow(
                <Counter initialValue={10} onChange={handleClick} />
            );

            wrapper.find('button').at(1).simulate('click');
            expect(handleClick).toHaveBeenCalledWith(11);
        });
    });

    describe('handlers', () => {
        it('calls onChange for input', () => {
            const handleChange = jest.fn();

            const wrapper = shallow(
                <Counter onChange={handleChange} />
            );

            wrapper.find('input').simulate('change', { target: { value: 22 } });
            expect(handleChange).toHaveBeenCalledWith(22);
        });
    });

    describe('handlers', () => {
        it('calls onChange for input with wrong value', () => {
            const handleChange = jest.fn();

            const wrapper = shallow(
                <Counter initialValue={10} onChange={handleChange} />
            );

            wrapper.find('input').simulate('change', {target: { value: 'hello' } });
            expect(handleChange).not.toBeCalled();
        });
    });

    describe('handlers', () => {
        it('calls onClick for left button near min limit', () => {
            const handleClick = jest.fn();

            const wrapper = shallow(
                <Counter initialValue={3} min={3} onChange={handleClick} />
            );

            wrapper.find('button').first().simulate('click');
            expect(handleClick).not.toHaveBeenCalledWith(2);
        });
    });

    describe('handlers', () => {
        it('calls onClick for right button near max limit', () => {
            const handleClick = jest.fn();

            const wrapper = shallow(
                <Counter initialValue={33} max={33} onChange={handleClick} />
            );

            wrapper.find('button').at(1).simulate('click');
            expect(handleClick).not.toHaveBeenCalledWith(34);
        });
    });
});
