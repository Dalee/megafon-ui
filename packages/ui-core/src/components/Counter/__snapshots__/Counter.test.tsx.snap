// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Counter /> layout renders Counter 1`] = `
<div
  className="mfui-beta-counter"
>
  <button
    className="mfui-beta-counter__btn mfui-beta-counter__btn_left"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    <span
      className="mfui-beta-counter__minus"
    >
      –
    </span>
  </button>
  <div
    className="mfui-beta-counter__input-box"
  >
    <input
      className="mfui-beta-counter__input"
      disabled={false}
      onBlur={[Function]}
      onChange={[Function]}
      value={0}
    />
  </div>
  <button
    className="mfui-beta-counter__btn mfui-beta-counter__btn_right"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <span
      className="mfui-beta-counter__plus"
    >
      +
    </span>
  </button>
</div>
`;

exports[`<Counter /> layout renders Counter with props 1`] = `
<div
  className="mfui-beta-counter mfui-beta-counter_disabled test-class"
>
  <button
    className="mfui-beta-counter__btn mfui-beta-counter__btn_left"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    <span
      className="mfui-beta-counter__minus"
    >
      –
    </span>
  </button>
  <div
    className="mfui-beta-counter__input-box"
  >
    <input
      className="mfui-beta-counter__input"
      disabled={true}
      onBlur={[Function]}
      onChange={[Function]}
      value={10}
    />
  </div>
  <button
    className="mfui-beta-counter__btn mfui-beta-counter__btn_right"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    <span
      className="mfui-beta-counter__plus"
    >
      +
    </span>
  </button>
</div>
`;
